File: src/test/java/mediathek/daten/DatenFilmGeoBlockingTest.java
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.daten;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class DatenFilmGeoBlockingTest {

    @Test
    void filmWithoutCountriesIsNotGeoBlocked() {
        var film = new DatenFilm();

        assertFalse(film.isGeoBlockedForLocation(Country.DE));
    }

    @Test
    void euFilmIsNotBlockedForEuLocation() {
        var film = new DatenFilm();
        film.addCountry(Country.EU);

        assertFalse(film.isGeoBlockedForLocation(Country.DE));
    }

    @Test
    void euFilmIsBlockedOutsideEu() {
        var film = new DatenFilm();
        film.addCountry(Country.EU);

        assertTrue(film.isGeoBlockedForLocation(Country.OTHER));
    }

    @Test
    void countrySpecificFilmIsBlockedOutsideAllowedCountry() {
        var film = new DatenFilm();
        film.addCountry(Country.DE);

        assertFalse(film.isGeoBlockedForLocation(Country.DE));
        assertTrue(film.isGeoBlockedForLocation(Country.AT));
    }
}

File: src/test/kotlin/mediathek/daten/DatenFilmGeoBlockingTest.kt
package mediathek.daten

import mediathek.config.application.ApplicationConfiguration
import mediathek.tool.FileSize
import okhttp3.HttpUrl.Companion.toHttpUrl
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

internal class DatenFilmGeoBlockingTest {

    @Test
    fun filmWithoutCountriesIsNotGeoBlocked() {
        val film = DatenFilm()

        assertFalse(film.isGeoBlockedForLocation(Country.DE))
    }

    @Test
    fun euFilmIsNotBlockedForEuLocation() {
        val film = DatenFilm()
        film.addCountry(Country.EU)

        assertFalse(film.isGeoBlockedForLocation(Country.DE))
    }

    @Test
    fun euFilmIsBlockedOutsideEu() {
        val film = DatenFilm()
        film.addCountry(Country.EU)

        assertTrue(film.isGeoBlockedForLocation(Country.OTHER))
    }

    @Test
    fun countrySpecificFilmIsBlockedOutsideAllowedCountry() {
        val film = DatenFilm()
        film.addCountry(Country.DE)

        assertFalse(film.isGeoBlockedForLocation(Country.DE))
        assertTrue(film.isGeoBlockedForLocation(Country.AT))
    }

    @Test
    fun forbiddenHlsSizeLookupMarksConfiguredCountryBlocked() {
        val previousLocation = ApplicationConfiguration.getInstance().geographicLocation
        ApplicationConfiguration.getInstance().geographicLocation = Country.AT
        try {
            val film = DatenFilm()
            val url = "https://example.org/video/master.m3u8"

            film.applyFileSizeLookupResult(
                url,
                FileSize.LookupResult(
                    byteLength = FileSize.INVALID_SIZE.toLong(),
                    httpStatusCode = 403,
                    resolutionUrl = "https://example.org/video/chunklist.m3u8".toHttpUrl(),
                    quality = FilmResolution.Enum.HIGH_QUALITY.name,
                ),
            )

            assertFalse(film.hasCountries())
            assertFalse(film.isGeoBlockedForLocation(Country.DE))
            assertTrue(film.isGeoBlockedForLocation(Country.AT))
            assertEquals(
                403,
                film.lookupFileSizeForUrl(url, false, FilmResolution.Enum.HIGH_QUALITY.name).httpStatusCode,
            )
        } finally {
            ApplicationConfiguration.getInstance().geographicLocation = previousLocation
        }
    }
}
